client/hwlib/src/collectors/cpuid.rs

```diff
@@ -0,0 +1,121 @@
+/* Copyright 2024 Canonical Ltd.
+ * All rights reserved.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+use smbioslib;
+
+pub(super) fn get_cpuid(
+    proc_info: &smbioslib::SMBiosProcessorInformation,
+) -> Result<String, Box<dyn std::error::Error>> {
+    let cpu_id = proc_info
+        .processor_id()
+        .map(|id| format!("{:x?}", id))
+        .ok_or("Processor ID not found")?;
+    Ok(cpu_id)
+}
+
+pub(super) fn convert_cpu_codename(cpu_id: &str) -> Result<String, Box<dyn std::error::Error>> {
+    if cpu_id.is_empty() {
+        return Ok(String::from("Unknown"));
+    }
+
+    let cpu_id_bytes: Vec<&str> = cpu_id.split_whitespace().take(3).collect();
+    if cpu_id_bytes.len() < 3 {
+        return Ok(String::from("Unknown"));
+    }
```

Comment:
In the logic above we take only 3 elements

Reply:
We take the first three elements, but is it okay if there are more which we ignore?

Reply:
Ah, something I meant to mention in my original comment, I think it's possible to avoid the allocation here by collecting into a tuple using [`.collect_tuple`](https://docs.rs/itertools/0.10.0/itertools/trait.Itertools.html#method.collect_tuple) provided by `itertools::Itertools` (`itertools` is a good crate)

Reply:
Ah, something I meant to mention in my original comment, I think it's possible to avoid the allocation here by collecting into a tuple using [`.collect_tuple`](https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.collect_tuple) provided by `itertools::Itertools` (`itertools` is a good crate)

Reply:
> is it okay if there are more which we ignore

yes, this is how the matching logic works even in the original code

Reply:
Updated in c5cfaa2